### ISSUE TITLE
How can client tell when `bind` is done?

### ISSUE THREAD

**Issue by jml** [warner/magic-wormhole#261]
server-protocol.md says:

> The first thing each client sends to the server, immediately after the WebSocket connection is established, is a `bind` message.

This message is intended to set state for the rest of the commands.

However, there's no response to the `bind` command, apart from `ack`, but…

> `ack` … is immediately sent back to the client for all commands

Which I take to mean that it's sent *before* the server handles the request.

How does the client know that it's safe to proceed with scoped commands?

**Comment by warner** [comment]
`ack` is just for the timing/debugging tools. Scoped commands are always safe to do after sending a `bind`, because they'll always arrive at the server after the `bind` (the websocket provides in-order reliable delivery of messages, and the connect-via-websocket state machines always sends a `bind` before announcing the connection to any other state machines, so there's no opportunity to send anything *before* the `bind`).

Does that seem reasonable?

**Comment by jml** [comment]
It seems reasonable, although I'd personally prefer something like an HTTP 204 on successful bind. I can imagine alternate server implementations where `bind` might fail (invalid app id; invalid side; server too busy, changed its mind since sending `welcome`).

Feel free to close.

**Comment by jml** [comment]
Thinking about this a little.

It seems a little inconsistent that `released` and `closed` are returned but that there's no `bound` (or `opened` for that matter). The inconsistency doesn't make sense at the protocol level, and only makes sense together with certain decisions about client implementation. 

I think having the protocol be internally consistent is a good thing (it's easier to understand, and makes for fewer special cases in code).

**Comment by jml** [comment]
Just bumped into an issue that is closely related to this and #260 but not exactly the same. Possibly deserves its own  ticket.

If a client does `open` twice, the server will send an error. A human can look at this error and tell that the second `open` failed, but AFAICT, there's no way for a computer to tell that.

**Comment by warner** [comment]
Yeah, please file a ticket for that. Messages are supposed to be idempotent. The client won't send the same message twice on (what it thinks is) the same connection, but if possible it'd be nice for the server to tolerate that. The thing about `open` is that it's a subscription request, and we're currently limiting clients to subscribing to a single mailbox per connection, so in this case idempotence might not be trivial.

**Comment by warner** [comment]
So I guess the way I think about `bind` is that it's amortizing the cost of sending `appid` and `side` with every subsequent message: if we wanted to enable sharing a single websocket connection between multiple clients in the same process (probably sending to distinct recipients), then we'd add `appid` and `side` in open/list/allocate/add/etc, and remove `bind` entirely. I don't want to enable that yet, but the protocol ought to behave the same either way. That's why I don't think `bind` needs a response.

That said, I also have in mind a size-limiting scheme for non-file-transfer applications (a Petmail addressbook+key -exchange is my typical example) where the clients might want to send more data than normal, and there needs to be some sort of negotiation between client and server about how much can be sent (how many messages, maximum size, how long they persist without at least one client keeping them alive, how long the mailbox can last despite keepalives). If the client asks for more than the server is willing to offer, there needs to be a response. The client's request could go into the `bind` along with the `appid`, and the server could send an `error` if the client asks for too much. Or the size/age request could go into the `open` or `allocate` instead.

So I still think that means `bind` doesn't need a response, but I'll admit it isn't completely clear-cut.
